Add store credits

JDutil opened the pull request:
This is an updated branch in favor of https://github.com/spree/spree/pull/5625
- [x] Frontend UX
- [x] Frontend UX Feature Specs
- [x] Backend UX for creating credit
- [x] Backend UX for creating credit reasons
- [ ] Final code review & houndci cleanup.

peterberkenbosch committed:
created the core models for store_credits

Commit:
Fix user class reference.

peterberkenbosch committed:
add the configurations for store credits

peterberkenbosch committed:
add factory for store_credit_events

peterberkenbosch committed:
use the correct configuration setting

peterberkenbosch committed:
add the store_credit_event model and migration

peterberkenbosch committed:
add the locale for amount_used in store_credits

peterberkenbosch committed:
fix the store_credit_spec, using the new rspec3 syntax

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://add-store-credits.spree-spree-9ed724b77adec04c49c2.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://add-store-credits.spree-spree-9ed724b77adec04c49c2.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

peterberkenbosch committed:
address hound comments

peterberkenbosch committed:
address hound comments

peterberkenbosch committed:
hound comments fixed

peterberkenbosch committed:
add store_credit payment method
add payment_method/store_credit and store_credit_event specs.

peterberkenbosch committed:
use do..end for multi-line block

peterberkenbosch committed:
Merge branch 'master' into add-store-credits

forvalho commented on core/app/models/spree/store_credit.rb:
```diff
@@ -0,0 +1,254 @@
+    def store_event
+      return unless amount_changed? || amount_used_changed? || amount_authorized_changed? || action == ELIGIBLE_ACTION
+
+      event = nil
```
Would'nt it make more sense like this?

```
event = if action
          store_credit_events.build(action: action)
        else
          store_credit_events.where(action: ALLOCATION_ACTION).first_or_initialize
        end
```

MattDunbar commented:
I feel like this would be better suited for its own gem.

Across many Spree builds we've done as an agency I'm yet to come across one that actually needs this, and building payment methods into the core seems a bit much while trying to keep it lightweight.

Although I'd lean towards this being an entirely separate gem in a separate repo, if its agreed that this needs to be part of the base it'd be great to isolate this and the existing check payment method into their own gem as part of this repo still.

athal7 commented:
@MattDunbar it already is an extension on spree-contrib. It was decided that the store-credit part of that extension would be useful for core, and the virtual gift card part of that extension is being left off for now until there appears to be more desire.

peterberkenbosch committed:
fixing some failing specs

peterberkenbosch committed:
Merge remote-tracking branch 'origin/master' into add-store-credits

Conflicts:
	core/app/models/spree/order.rb

peterberkenbosch committed:
no more environment column for payment_methods

peterberkenbosch commented:
@MattDunbar @athal7 while trying to get this merged into core, I tend to feel the same Matt. It's actually a lot of code, and also has some limitations currently (only works with a creditcard payment method). Will discuss this further today. @JDutil wdyt.

peterberkenbosch committed:
move over the decorations to the proper models.

peterberkenbosch committed:
Merge remote-tracking branch 'upstream/master' into add-store-credits

athal7 commented:
You can remove the credit card restriction, that is Bonobos logic creeping in. Let me know if a chat would be beneficial or if you want me to take a look at certain areas that are concerning to you.

peterberkenbosch commented:
Yes let's chat!

peterberkenbosch committed:
add the order/store_credit specs (wip)
remove environment from payment method factory
remove bonobos logic from order.rb

peterberkenbosch committed:
adding specs for store_credits on payments and reimbursement_type and users

peterberkenbosch committed:
no super needed call

paultyng commented:
This still has the issue of depending on presence of a confirm step.  You can see our workarounds in [my comment in the previous issue](https://github.com/spree/spree/pull/5625#issuecomment-69361879).  Besides removing dependency on the optional step though, I think our changes simplified the addition of the payments by removing the need to do the reconciliation with existing payments.

I'd be happy to chat in the gitter room about our workarounds we've implemented as well to get the extension on our prod site if you are interested, I wish I had some time to contribute on this one as I'd really like to get to stop maintaining this part of our codebase, but I probably won't free up for a couple weeks yet, if then.

peterberkenbosch committed:
Merge remote-tracking branch 'upstream/master' into add-store-credits

peterberkenbosch committed:
Merge remote-tracking branch 'upstream/master' into add-store-credits

peterberkenbosch committed:
Merge remote-tracking branch 'upstream/master' into add-store-credits

peterberkenbosch committed:
add locales for store_credit_payment_method

peterberkenbosch committed:
Merge remote-tracking branch 'upstream/master' into add-store-credits

peterberkenbosch committed:
Merge remote-tracking branch 'upstream/master' into add-store-credits

peterberkenbosch committed:
fixing all errors for the order/store_credit_specs, still failures left.

peterberkenbosch committed:
Merge remote-tracking branch 'upstream/master' into add-store-credits

peterberkenbosch committed:
fix specs on store credits reimbursement specs

peterberkenbosch committed:
fix user specs with store credits

peterberkenbosch committed:
remove the display_event_date, using to_date like everywhere else

peterberkenbosch committed:
allow nil for delegating store_credit? to PaymentMethod

peterberkenbosch committed:
add stub for payments.store_credits.pending fixing the state_machine specs

peterberkenbosch committed:
store credits is a payment method by default present.

peterberkenbosch committed:
fix the payment method counts since the store credit payment method is by default always present.

peterberkenbosch committed:
refactor the event assignment to be a bit more clean

peterberkenbosch committed:
add the reconcile credit card method back for now to get refactoring starten with passing specs.

peterberkenbosch committed:
fix store_credit payment specs.

peterberkenbosch committed:
Merge remote-tracking branch 'upstream/master' into add-store-credits

peterberkenbosch committed:
Merge remote-tracking branch 'upstream/master' into add-store-credits

peterberkenbosch committed:
get rid of the reconcile creditcard method and use the before payment state to add any store credits.

peterberkenbosch committed:
Merge remote-tracking branch 'upstream/master' into add-store-credits

peterberkenbosch committed:
Merge remote-tracking branch 'upstream/master' into add-store-credits

peterberkenbosch committed:
set expectation on any_instance not on the class level.

peterberkenbosch committed:
Merge remote-tracking branch 'upstream/master' into add-store-credits

peterberkenbosch committed:
more passing specs.

peterberkenbosch committed:
fix some more specs

peterberkenbosch committed:
final failing spec?

peterberkenbosch committed:
Merge remote-tracking branch 'upstream/master' into add-store-credits

nimashariatian commented:
just curious if this is going to be merged soon?

peterberkenbosch commented:
Yes it will. Some minor issues to finish.

On Mon, Mar 2, 2015 at 11:10 PM, nimashariatian <notifications@github.com>
wrote:

> just curious if this is going to be merged soon?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spree/spree/pull/5917#issuecomment-76837422>.
>



-- 

Peter Berkenbosch
Spree Commerce, Inc.

iangreenleaf commented on core/app/models/spree/store_credit.rb:
```diff
@@ -0,0 +1,253 @@
+      if amount_used > 0
+        errors.add(:amount_used, 'is greater than zero. Can not delete store credit')
+      end
+    end
```
This method should return false, or else the destroy will go ahead. You could do it like this:

```ruby
def validate_no_amount_used
  if amount_used > 0
    errors.add(...')
  end
  errors.blank?
end
```

JDutil replied:
Feel free to submit a PR against the store credit branch to fix this.

smartacus committed:
Fixed null timestamp deprecation warnings

Commit:
Merge branch 'master' into add-store-credits

smartacus committed:
Fix hound issues

smartacus committed:
Fix Store Credit validation errors

smartacus committed:
Removed unused method from Order

smartacus committed:
Implement interface for applying store credits to users

smartacus committed:
Fix some hound issues

smartacus committed:
Move permitted_attributes from store_credits_controller

smartacus committed:
Add spree prefix to routes

smartacus committed:
Implement admin interface for store credit categories

smartacus committed:
Fix broken spec

smartacus committed:
Move store credit methods out of order and into module

JDutil committed:
Merge pull request #6634 from TrialGuides/add-store-credits

Updates to Store Credits

mleglise committed:
New methods: Order.could_use_store_credit? and .using_store_credit?

mleglise committed:
Added checkout#payment view for Store Credit.

mleglise committed:
New method: Order.total_applied_store_credit

returns only the store credit payments that have actually been applied to this order.

mleglise committed:
Display applied store credit in checkout order summary.

- new method: Order.display_total_applied_store_credit
- DRY Order.total_applicable_store_credit

mleglise committed:
Checkout#payment: display the remaining order total after store credit.

mleglise committed:
Added Spree::PaymentMethod::StoreCredit to config.spree.payment_methods

Fixes editing the payment method via the backend.

mleglise committed:
Display payment amount next to method name.

Primarily affects the Order status page, post-purchase.

mleglise committed:
Do not automatically add Store Credit payments to an order.

Removing this from the default checkout flow to enable user choice, similar to Amazon.

mleglise committed:
Checkout#payment: display store credit prompt above the normal payment methods.

Expects that StoreCredit Payment Method will have `display_on = “back_end”`.

mleglise committed:
Spree.disableSaveOnClick: do not truly disable the submit buttons, so that their name gets sent as a param.

Required for the “Apply Store Credit” opt-in button.

mleglise committed:
Checkout#payment: trigger add_store_credit_payments if user clicked “Apply Store Credit” button.

mleglise committed:
Merge remote-tracking branch 'upstream/master' into add-store-credits

Conflicts:
	core/app/models/spree/order.rb
	core/app/models/spree/payment.rb
	core/config/initializers/user_class_extensions.rb
	core/spec/models/spree/order/checkout_spec.rb
	core/spec/models/spree/payment_method_spec.rb
	frontend/app/controllers/spree/checkout_controller.rb

mleglise committed:
Fixed broken order spec.

Temporarily commented out payment.cancel! method. Need to evaluate best course of action.

mleglise committed:
Admin::Payments#new: added source form for StoreCredit.

mleglise committed:
Added data-hook attrs to Store Credit payment views.

mleglise committed:
Fix: 2nd Payment Method should only be charged for the remaining total.

mleglise committed:
Removed / updated old Store Credit specs, dealing with outdated functionality.

mleglise committed:
Fixed spec: checkout prevents double form submit differently now.

The submit buttons are only soft-disabled. The JS handler function itself prevents the double submit.

mleglise committed:
Admin::Users: add store credit balance to lifetime stats.

mleglise committed:
Admin::Payments#show: added source view for StoreCredit.

mleglise committed:
PaymentMethod::StoreCredit.cancel: updated to conform with standard payment method action syntax.

JDutil committed:
Merge pull request #6647 from mleglise/add-store-credits

Make store credits optional, instead of automatic

johnwook commented:
Thanks for the great works, all! I think merging is under pending because of some conflicts. Anyone that can handle merge conflicts, or is there anything that I can do to make this happen? :) :+1:

JDutil commented:
@johnwook you could submit a PR against add-store-credits branch that merges master branch into it, and/or improve/review/test the PR descriptions task list.

jasonfb commented on backend/app/controllers/spree/admin/store_credits_controller.rb:
```diff
@@ -0,0 +1,89 @@
+        @store_credit = Spree::StoreCredit.find_by_id(params[:id]) || Spree::StoreCredit.new
+      end
+
+      def ensure_unused_store_credit
```
two things: 1) this is domain logic and shouldn't be in a controller. (But that is forgivable)

2) We actually have a specific need from our Customer Service team that store credits be adjustable after they are created. In our implementation, we enforce that you cannot lower the store credit amount to below the used + authorized amounts.

jasonfb replied:
also... raising errors is no good user experience for my Admins. Better to show them pretty errors, regardless of how we go with the functionality.

jasonfb commented on core/app/models/spree/payment.rb:
```diff
@@ -30,6 +30,7 @@ class Payment < Spree::Base
 
     # invalidate previously entered payments
     after_create :invalidate_old_payments
+    after_create :create_eligible_credit_event
```
is there a way we can do this without callbacks? Overuse of callbacks is large problem in Spree, and should be curtailed

jasonfb commented:
@johnwook @JDutil 

At Mack Weldon we have a compelling need to get this into Spree 3.1 (we are currently using our own SC implementation and I would like to merge it with this implementation when we upgrade), so I can probably put some dev resources on this within the next 3 week timeframe. 

I left some minor feedback ... since so much work has gone into this already, I would not consider my feedback to be show-stopper for this moving forward as-is (since 1, the feedback is minor, and 2, those things could be addressed in subsequent releases)

Perhaps all it needs to be squashed & rebased?

jasonfb commented:
The only other 'feature' we need is to be able to expire store credits--- this is *very* important to our stakeholders, as it affects the amount of debt on our books. That, too, is a feature which I think can be added later, and one in which I can assign some developer resources to building out.

JDutil commented:
@jasonfb I don't think a squash rebase is necessary this is pretty large branch, and I'm fine with it just being merged to maintain history.

I believe there was the ability to expire store credits in this PR already, but they are non-expiring by default. It's been awhile since I've reviewed this though.

jasonfb commented:
Re: Git Process.. I will defer to you (in my app I always squash & rebase because I hate merge commits). 

Re: Expiration....  Unless I've missed something, assuming this implementation is based on bonobos/spree-store-credit-as-payment-method (which it looks like, with some small differences), confusingly, there is an "expiring" and "non-expiring" *type* --- and there is some calculus to apply the expiring ones before the non-expiring ones, but the *actual expiration* is not implemented. in other words, some of the code is written anticipation of expiration, but actual expiration is not implemented (this is also true for bonobos/spree-store-credit-as-payment-method, in which the author stated simply that the expiration was a planned future extension)

You can see I'm right just but looking at the store credits table itself --- it has no expires_at field.

JDutil commented:
Usually I prefer the squash/rebase, but with large PRs like this it's best to keep the history.

This implementation was based on the bonobos extension, but has had a different frontend for it implemented as they just used the API. 

They may have finished working on expiration in solidus or maybe not I'm not really sure.

johnwook commented:
@JDutil @jasonfb Thanks for replies! I'll try my best to make this branch merged. Also, I'm looking forward to seeing @jasonfb 's pre-implemented store credit functions including expiration.

jasonfb commented:
@johnwook --- in our app (which uses similar code lifted from bonobos/store_credit_as_payment_method) we did the expiration in a really wonky way that I'm not particularly happy with. It basically runs a cron job and on the day of expiration, reset the ```amount``` of the store credit to be whatever you haven't yet used (either 0 or the amount used+amount authorized)

it's not very elegant, so I wouldn't even recommend it for the core.

justindrees commented:
@johnwook would love to have this functionality for a site I'm working on. Any update on when/if this will be merged? Many thanks!

johnwook commented:
Hi @justindrees, sorry for late reply.
Recently, I moved from srpee to [solidus](solidus.io), and solidus has built in store-credit functionality.
Check some articles why I moved to solidus.

- [The Future of Spree Open Source Software](https://spreecommerce.com/blog/future-of-spree-oss)
- [The Future of Spree and the Future of Solidus](https://solidus.io/blog/2015/10/28/future-of-spree.html)

It took a couple of days for me to complete the migration(spree 3.0 branch -> solidus)

Mafi88 committed:
Merge branch 'master' into add-store-credits

Conflicts:
	core/app/models/spree/payment.rb
	core/config/initializers/user_class_extensions.rb

Mafi88 committed:
Frontend UX feature specs

Fix houndci/codeclimate issues for user_spec

Fix houndci issues for store_credit

Fix houndci issues for store_credit_spec

Fix more houndci issues

Finish payment canceling

Set StoreCredit PaymentMethod display_on to back_end

Fix more houndci issues

Mafi88 commented:
@damianlegawiec @priyank-gupta Could you please check if everything is ok here?

Don't correct code according to: ```Use && instead of and```. There should be ```and``` as it is flow operator, ```&&``` will break changing path.

damianlegawiec committed:
Merge branch 'master' into add-store-credits

damianlegawiec committed:
store credits admin: before_filter -> before_action

damianlegawiec committed:
store credit model: using associations for validation rather than columns

damianlegawiec committed:
store credits: small refactorings & rubocop fixes

damianlegawiec committed:
store credits: better handling of store credits admin page for non-existend user
